### PR TITLE
importccl: allow clients to disconnect and jobs to still finish

### PR DESCRIPTION
**Pull request by ajwerner**
This commit deals with the case where the client disconnects. Prior to this
commit, the job would attempt to send its results on the client resultsCh
and block forever.

Release justification: fixes for high-priority or high-severity bugs in existing functionality

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
I imagine we'll need to do this for backups too, just want to validate the approach.

File: pkg/ccl/importccl/import_stmt.go [hunk 1]
```diff
@@ -685,11 +690,45 @@ func importPlanHook(
...
+	resultsCh chan<- tree.Datums,
+	resultsFromJob <-chan tree.Datums,
+) error {
+	readResultsFromJob := func() {
+		for range resultsFromJob {
+		}
+	}
+	for {
+		select {
+		case r := <-resultsFromJob:
```
Comment (dt) [inline pkg/ccl/importccl/import_stmt.go]: if `resultsFromJob`  is closed r is nil, right? will this just inf then? loop? should we check `errCh` first?

File: pkg/ccl/importccl/import_stmt.go [hunk 2]
```diff
@@ -685,11 +690,45 @@ func importPlanHook(
...
+			select {
+			case <-ctx.Done():
```
Comment (dt) [inline pkg/ccl/importccl/import_stmt.go]: should we also check errCh here?

File: pkg/ccl/importccl/import_stmt.go [hunk 3]
```diff
@@ -685,11 +690,45 @@ func importPlanHook(
...
 	return fn, backupccl.RestoreHeader, nil, false, nil
 }
 
+// copyToResultsChannel exists to deal with the fact that the resultsCh hooked
+// up to the conn executor does not notify the plan hook when the connection
+// goes away. When that connection goes away however, the context will be
+// canceled. Because the started job needs to send on the resultsFromJobs
+// channel in order to complete, this function launches a goroutine to consume
+// from that channel in the case of context cancellation.
+func copyToResultsChannel(
```
Comment (dt) [inline pkg/ccl/importccl/import_stmt.go]: I think this helper should live in the `jobs` package ? assuming everyone using resultsCh has the same concern.

**Review by dt — APPROVED** [review]

**Review by ajwerner — COMMENTED** [review]
Okay, moved the logic to jobs. RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-:-M2KrkQd1IUeK7B2RriQ:b-hby7il)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/importccl/import_stmt.go, line 704 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2Jy6S5bbk1e5oW8q5n-r1-704:-M2K3DIDAMu7Uy8nfBxh:b-ncix83) ([raw file](https://github.com/cockroachdb/cockroach/blob/bea24f99102ab5475ecc4c4cd2bf38bdb574f3b7/pkg/ccl/importccl/import_stmt.go#L704)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think this helper should live in the `jobs` package ? assuming everyone using resultsCh has the same concern.
</blockquote></details>

make it a method on `StartableJob`. I think it's cleaner this way.
___
*[pkg/ccl/importccl/import_stmt.go, line 716 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2Jy6S5bbk1e5oW8q5n-r1-716:-M2Krf9qCf1PqsHVMVWL:bilct95) ([raw file](https://github.com/cockroachdb/cockroach/blob/bea24f99102ab5475ecc4c4cd2bf38bdb574f3b7/pkg/ccl/importccl/import_stmt.go#L716)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

if `resultsFromJob`  is closed r is nil, right? will this just inf then? loop? should we check `errCh` first?
</blockquote></details>

dealt with this case.
___
*[pkg/ccl/importccl/import_stmt.go, line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2Jy6S5bbk1e5oW8q5n-r1-718:-M2K3fdb1Kg1WHga_OGr:b7tsht1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bea24f99102ab5475ecc4c4cd2bf38bdb574f3b7/pkg/ccl/importccl/import_stmt.go#L718)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

should we also check errCh here?
</blockquote></details>

I flattened the loop to make reduce the number of cases.

**Review by dt — APPROVED** [review]
> Reviewed 2 of 6 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46089)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/ccl/importccl/import_stmt.go, line 704 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2Jy6S5bbk1e5oW8q5n-r1-704:-M2M5ses5fGoJl7bvBra:b-1hf9xe) ([raw file](https://github.com/cockroachdb/cockroach/blob/bea24f99102ab5475ecc4c4cd2bf38bdb574f3b7/pkg/ccl/importccl/import_stmt.go#L704)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

make it a method on `StartableJob`. I think it's cleaner this way.
</blockquote></details>

yeah, this is much cleaner!
___
*[pkg/ccl/importccl/import_stmt.go, line 663 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2M5y2CCNO_T-uB8fjx:-M2M5y2CCNO_T-uB8fjy:b-xpu2k3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e243f6d6f17f6b3f6781996d2a980aa30cf9710/pkg/ccl/importccl/import_stmt.go#L663)):*
> ```Go
> 		// with the client disconnecting. If we did not use this intermediate
> 		// channel, we could leave the job blocked forever trying to send results.
> 		// See copyToResultChannel().
> ```

I think this comment is now out of date / unneeded.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 3834 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2M8ToODfE71yQ-4s5x:-M2M8ToODfE71yQ-4s5y:bqylkz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e243f6d6f17f6b3f6781996d2a980aa30cf9710/pkg/ccl/importccl/import_stmt_test.go#L3834)):*
> ```Go
> 
> 	// Launch a subprocess ot kick off the IMPORT.
> 	cmd := exec.Command(os.Args[0], "--test.run", "TestImportClientDisconnectHelper",
> ```

we need a subprocess for this? we can't just like cancel its context or something?

**Review by ajwerner — COMMENTED** [review]
TFTR! Addressed the remaining comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-:-M2YnEtk86KmBQkUsTiF:bugymfi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[pkg/ccl/importccl/import_stmt.go, line 663 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2M5y2CCNO_T-uB8fjx:-M2YnEMO5vLN2cMO-ZAj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e243f6d6f17f6b3f6781996d2a980aa30cf9710/pkg/ccl/importccl/import_stmt.go#L663)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think this comment is now out of date / unneeded.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 3834 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46089#-M2M8ToODfE71yQ-4s5x:-M2YmpF3-EjatjkN9Lwn:b-phpk29) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e243f6d6f17f6b3f6781996d2a980aa30cf9710/pkg/ccl/importccl/import_stmt_test.go#L3834)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

we need a subprocess for this? we can't just like cancel its context or something?
</blockquote></details>

No, just needed to use pgx and not pq. Fixed.

**Review by dt — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46089)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)

**Comment by ajwerner** [comment]
TeamCity failed in some JS failure, seems to have passed on re-run. 

bors r=dt

**Pull request merged**